From d4921ccf5f61be6324bba3b14b25f29e28c5b2bd Mon Sep 17 00:00:00 2001
From: Nicholas Guriev <nicholas@guriev.su>
Date: Wed, 20 Apr 2022 12:34:08 +0300
Subject: [PATCH] WaylandProtocols for Qt5

Also set the wayland_protocols_loc variable to prefer XML definitions from a
system-wide package if installed.
---
 CMakeLists.txt | 23 +++++++++++++++++++++--
 1 file changed, 21 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index bb67d85..6e94c88 100644
--- a/Telegram/lib_ui/CMakeLists.txt
+++ b/Telegram/lib_ui/CMakeLists.txt
@@ -4,6 +4,13 @@
 # For license and copyright information please follow this link:
 # https://github.com/desktop-app/legal/blob/master/LEGAL
 
+if (NOT DESKTOP_APP_DISABLE_WAYLAND_INTEGRATION AND NOT DESKTOP_APP_QT6)
+    find_package(ECM REQUIRED)
+    set(CMAKE_MODULE_PATH ${ECM_MODULE_PATH})
+    find_package(QtWaylandScanner REQUIRED)
+    find_package(WaylandProtocols)
+endif()
+
 add_library(lib_ui STATIC)
 add_library(desktop-app::lib_ui ALIAS lib_ui)
 init_target(lib_ui)
@@ -20,6 +27,12 @@ set(style_files
     ui/widgets/widgets.style
 )
 
+if (WaylandProtocols_FOUND)
+    set(wayland_protocols_loc ${WaylandProtocols_DATADIR})
+else()
+    set(wayland_protocols_loc ${third_party_loc}/wayland-protocols)
+endif()
+
 generate_palette(lib_ui ui/colors.palette)
 generate_styles(lib_ui ${src_loc} "${style_files}" ui/colors.palette)
 generate_emoji(lib_ui emoji.txt emoji_suggestions/emoji_autocomplete.json)
@@ -273,10 +286,16 @@ elseif(LINUX)
     if (DESKTOP_APP_QT6)
         qt6_generate_wayland_protocol_client_sources(lib_ui
         FILES
-            ${third_party_loc}/wayland-protocols/stable/xdg-shell/xdg-shell.xml
+            ${wayland_protocols_loc}/stable/xdg-shell/xdg-shell.xml
         )
     else()
-        message(FATAL_ERROR "This piece of cmake code is not ported to Qt 5")
+        ecm_add_qtwayland_client_protocol(xdg_shell_protocol_srcs
+            PROTOCOL ${wayland_protocols_loc}/stable/xdg-shell/xdg-shell.xml
+            BASENAME xdg-shell
+        )
+
+        target_sources(lib_ui PRIVATE ${xdg_shell_protocol_srcs})
+        target_include_directories(lib_ui PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
     endif()
 
     target_link_libraries(lib_ui
